### PR TITLE
Transpose and Evert

### LINKED ISSUE #29: Consider function to invert Option<Result<,>> to Result<Option<>,>

### PR DESCRIPTION
**Pull request by RyanMarcotte**
Closes #29 

I added both [`Transpose`](https://en.wiktionary.org/wiki/transpose#English) ("swap; interchange") and [`Evert`](https://en.wiktionary.org/wiki/evert#English) ("turn inside out") extension methods.
- `Transpose` takes `Result<T1, T2>` as input and produces `Result<T2, T1>`
- `Evert` takes `Option<Result<TSuccess, TFailure>>` as input and produces `Result<Option<TSuccess>, TFailure>`

~The names are different so that it is easy to disambiguate between the two methods; there may have been conflicts otherwise since the function signatures are similar.  I don't see `Evert` being a common use case, so I am OK with a lesser-known name.~  The two methods operate on different types, so there would be no conflicts.  I do like the separate names though.

### CODE REVIEW

**Comment by RyanMarcotte** [comment]
TLDR: C#'s type system sometimes requires a bunch of typing when using `Result<TSuccess, TFailure>`

I've only begun experimenting with this, so YMMV, but some validation rule chains can be made less verbose by producing an `Option<Error>` first, then using `ToResult(() => Unit.Value).Transpose()` to get `Result<Unit, Error>`.

``` csharp
// validation approach 1...  duplicate typing of Result<Unit, Error> across all method signatures in the chain
public Result<Thing, Error> MakeThing(Parameters parameters)
    => Result.Unit<Error>()
        .Bind(_ => ValidationRule1(parameters))
        .Bind(_ => ValidationRule2(parameters))
        .Bind(_ => ValidationRule3(parameters))
        .Map(_ => new Thing(...));

private Result<Unit, Error> ValidationRule1(Parameters parameters) => Result.Create(...);
private Result<Unit, Error> ValidationRule2(Parameters parameters) => Result.Create(...);
private Result<Unit, Error> ValidationRule3(Parameters parameters) => Result.Create(...);
```

``` csharp
// validation approach 2...  duplicate typing of Result<Unit, Error> across all return statements
public Result<Thing, Error> MakeThing(Parameters parameters)
{
    if (some condition 1) return Result.Failure<Thing, Error>(...);
    if (some condition 2) return Return.Failure<Thing, Error>(...);
    if (some condition 3) return Return.Failure<Thing, Error>(...);

    return Result.Success<Thing, Error>(new Thing(...));
}
```

``` csharp
// new validation approach...  fewer generics in validation logic, though chain is slightly more verbose and potentially less readable due to ToResult(...).Transpose()
public Result<Thing, Error> MakeThing(Parameters parameters)
    => MakeError(parameters)
        .ToResult(() => Unit.Value)
        .Transpose()
        .Map(_ => new Thing(...));

private Option<Error> MakeError(Parameters parameters)
{
    if (some condition 1) return Option.Some(...);
    if (some condition 2) return Option.Some(...);
    if (some condition 3) return Option.Some(...);

    return Option.None<Error>();
}
```

If there is a separate approach not covered by the above three snippets I am interested to see it.  I have been using the third approach to clean up [smart constructors](https://wiki.haskell.org/Smart_constructors) specifically.

The verbosity shown in the first two snippets would be alleviated [if we could define aliases globally to make reuse of particular `Result<TSuccess, TFailure>` types easier](https://github.com/dotnet/csharplang/issues/1170) (i.e. `public struct ThingFactoryResult aliasof Result<Thing, Error>`).  As you know, C# does not allow inheritance of `struct` types.  An alternative approach could be wrapping a particular `Result<TSuccess, TFailure>` type in a `class`, but that defeats the purpose of this library: avoiding allocations for raw performance gains.

**Comment by RyanMarcotte** [comment]
I thought about such an extension method too, but decided against adding it right away since it doesn't save that much typing.  We can do that as a separate PR if people would find it valuable.

Thanks for looking this over!
